frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx

```diff
@@ -611,22 +612,32 @@ const crdCardRowStateToProps = ({ k8s }, { crdDescs }) => {
   };
 };
 
-export const CRDCardRow = connect(crdCardRowStateToProps)((props: CRDCardRowProps) => (
-  <div className="co-crd-card-row">
-    {_.isEmpty(props.crdDescs) ? (
-      <span className="text-muted">No Kubernetes APIs are being provided by this Operator.</span>
-    ) : (
-      props.crdDescs.map((desc) => (
-        <CRDCard
-          key={referenceForProvidedAPI(desc)}
-          crd={desc}
-          csv={props.csv}
-          canCreate={props.createable.includes(referenceForProvidedAPI(desc))}
-        />
-      ))
-    )}
-  </div>
-));
+export const CRDCardRow = connect(crdCardRowStateToProps)((props: CRDCardRowProps) => {
+  const internals = getInternalObjects(props.csv);
+  return (
+    <div className="co-crd-card-row">
+      {_.isEmpty(props.crdDescs) ? (
+        <span className="text-muted">No Kubernetes APIs are being provided by this Operator.</span>
+      ) : (
+        props.crdDescs.reduce(
```

Comment:
It seems that `isInternalObject` function is used to gate the mapping from `props.crdDescs` to corresponding `<CRDCard>` components.

However, I think that plain `filter` + `map` is just better (simpler, more readable) than `reduce` here:

```ts
props.crdDescs
  .filter((desc) => !isInternalObject(internalObjects, desc.name))
  .map((desc) => (
    <CRDCard
      key={referenceForProvidedAPI(desc)}
      crd={desc}
      csv={props.csv}
      canCreate={props.createable.includes(referenceForProvidedAPI(desc))}
    />
  ));
```

Alternatively, you can skip the filter part and use `_.compact` to filter-out falsy (i.e. `null`) values from the array.

Reply:
I leaned towards filter, map combination at first, but then again the question would be why iterate over the same stuff twice?

Reply:
Yes, I tend to lean towards reduce when we have a few different operations to perform iteratively. If we get too comfortable with chaining iterative functions like this, it makes it easy to create performance issues.

Reply:
Personally, I find `filter` + `map` simpler because it separates the two logical operations and thus better conveys the intent of the code, while `reduce` combines them both into one.

I agree that `reduce` is more performant and unnecessary iterations should be avoided if possible. I'm OK with using `reduce` here.

Reply:
Old conversation, but I'm good with either approach. Both are linear time complexity, and these lists are never very large. I lean toward focusing on readability and then identifying bottlenecks later through profiling.

Having said that, `reduce` seems perfectly fine here.